cpp/src_prims/stats/histogram.h

```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
   CUDA_CHECK(cudaGetLastError());
 }
 
+DI void clearHashTable(int2* ht, int hashSize) {
+  for (auto i = threadIdx.x; i < hashSize; i += blockDim.x) {
+    ht[i] = {-1, 0};
+  }
+  __syncthreads();
+}
+
+DI int findEntry(int2* ht, int hashSize, int binId) {
+  int idx = binId % hashSize;
+  int t;
+  int count = 0;
+  while ((t = atomicCAS(&(ht[idx].x), -1, binId)) != -1 && t != binId) {
+    ++count;
+    if (count >= hashSize) {
+      idx = -1;
+      break;
+    }
+    ++idx;
+    if (idx >= hashSize) {
+      idx = 0;
+    }
+  }
+  return idx;
+}
+
+DI void flushHashTable(int2* ht, int hashSize, int* bins) {
+  __syncthreads();
+  for (auto i = threadIdx.x; i < hashSize; i += blockDim.x) {
+    if (ht[i].x != -1 && ht[i].y > 0) {
+      atomicAdd(bins + ht[i].x, ht[i].y);
+    }
+    ht[i] = {-1, 0};
+  }
+  __syncthreads();
+}
+
 template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
 __global__ void smemHashHistKernel(int* bins, const DataT* data, IdxT n,
                                    IdxT nbins, BinnerOp binner, int hashSize) {
-  //extern __shared__ int2 smem[];
-  ///@todo!
+  extern __shared__ int2 ht[];
+  int* needFlush = (int*)&(ht[hashSize]);
+  if (threadIdx.x == 0) {
+    needFlush[0] = 0;
+  }
+  clearHashTable(ht, hashSize);
+  IdxT tid = threadIdx.x + IdxT(blockDim.x) * blockIdx.x;
+  IdxT stride = IdxT(blockDim.x) * gridDim.x;
+  int nCeil = ceildiv<int>(n, stride);
+  for (auto i = tid; i < nCeil; i += stride) {
+    bool iNeedFlush = false;
+    int binId = 0;
+    if (i < n) {
+      binId = binner(data[i], i);
+      int hidx = findEntry(ht, hashSize, binId);
+      if (hidx >= 0) {
+        atomicAdd(&(ht[hidx].y), 1);
+      } else {
+        needFlush[0] = 1;
+        iNeedFlush = true;
+      }
+    }
+    __syncthreads();
+    if (needFlush[0]) {
+      flushHashTable(ht, hashSize, bins);
+      if (threadIdx.x == 0) {
+        needFlush[0] = 0;
+      }
+      __syncthreads();
+    }
+    if (iNeedFlush) {
+      int hidx = findEntry(ht, hashSize, binId);
+      atomicAdd(&(ht[hidx].y), 1);
+    }
+  }
+  flushHashTable(ht, hashSize, bins);
 }
 
 template <typename DataT, typename BinnerOp, typename IdxT, int TPB, int VecLen>
 void smemHashHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
                   cudaStream_t stream) {
   int nblks = ceildiv<int>(VecLen ? n / VecLen : n, TPB);
-  auto maxSmem = maxSharedMem();
-  auto maxHashEntries = int(maxSmem / sizeof(int2));
+  // NOTE: assumes 48kB smem!
+  int hashSize = 6047;
```

Comment:
You could make this work for a variety of shmem sizes with a cascading if structure.  Something like `if (maxSmem > 96k) hashSize = (some prime number bigger than 96k/8) else if (maxSmem > 64k) hashSize = (some prime number up to 96k) else if (maxSmem > 48k) ... else hashSize = 993;

Reply:
This is ugly.  I'm sure you can come up with a pretty solution based on a handful of smem sizes.

Reply:
Fair enough. It should be easy to do that automatically. Let me do it.

Reply:
Added a simple seive-based technique to automatically figure out the best hash table size for a given shared memory configuration.

Reply:
I didn't mean to suggest that you write an entire sieve.  I was thinking of something more like this:
auto maxSmem = maxSharedMem();
auto maxHashEntries = int(maxSmem / sizeof(int2));
int hashSize = 997;
if (maxHashEntries >= 16381) hashSize = 16381;
if (maxHashEntries >= 32749) hashSize = 32749;
if (maxHashEntries >= 49139) hashSize = 49139;
if (maxHashEntries >= 65521) hashSize = 65521;
if (maxHashEntries >= 98299) hashSize = 98299;
if (maxHashEntries >= 131071) hashSize = 131071;
if (maxHashEntries >= 196657) hashSize = 196657;

Reply:
Except I used the wrong values here.  I was thinking in bytes, not hash entries.

Reply:
I see! We'll need such a thing in future at other places too. So, since this is already written, I'll just keep this one. :) Thank you for catching this one.